packages/svelte/src/compiler/phases/2-analyze/css/css-prune.js

```diff
@@ -353,46 +348,35 @@ function relative_selector_might_apply_to_node(relative_selector, rule, element)
 					)
 				)
 			);
-		if (include_self) {
-			child_elements.push(element);
-			descendant_elements.push(element);
-		}
 
-		const seen = new Set();
+		// set them lazy because it's expensive to calculate
+		/** @type {Array<Compiler.AST.RegularElement | Compiler.AST.SvelteElement>} */
+		let descendant_elements;
+		/** @type {Array<Compiler.AST.RegularElement | Compiler.AST.SvelteElement>} */
+		let sibling_elements;
+		/** @type {Array<Compiler.AST.RegularElement | Compiler.AST.SvelteElement>} */
+		let sibling_descendant_elements;
 
 		/**
-		 * @param {Compiler.AST.SvelteNode} node
-		 * @param {{ is_child: boolean }} state
+		 * @param {ExtendedRelativeSelector[]} selectors
 		 */
-		function walk_children(node, state) {
-			walk(node, state, {
-				_(node, context) {
-					if (node.type === 'RegularElement' || node.type === 'SvelteElement') {
-						descendant_elements.push(node);
-
-						if (context.state.is_child) {
-							child_elements.push(node);
-							context.state.is_child = false;
-							context.next();
-							context.state.is_child = true;
-						} else {
-							context.next();
-						}
-					} else if (node.type === 'RenderTag') {
-						for (const snippet of node.metadata.snippets) {
-							if (seen.has(snippet)) continue;
+		const get_elements = (selectors) => {
+			const left_most_combinator = selectors[0]?.combinator ?? descendant_combinator;
 
-							seen.add(snippet);
-							walk_children(snippet.body, context.state);
-						}
-					} else {
-						context.next();
-					}
-				}
-			});
-		}
+			if (left_most_combinator.name === ' ' || left_most_combinator.name === '>') {
+				descendant_elements ??= get_descendant_elements(element, include_self);
+				return descendant_elements;
+			}
 
-		walk_children(element.fragment, { is_child: true });
+			sibling_elements ??= get_following_sibling_elements(element, include_self);
+			if (selectors.some((s) => s.combinator?.name === ' ' || s.combinator?.name === '>')) {
```

Comment:
It doesn't cover `s.combinator === null` case - I pose only the first selector can have no combinator.